File: src/435_non_overlapping_intervals/non_overlapping_intervals.cpp
#include <algorithm>
#include <iostream>
#include <std>
#include <vector>

class Solution {
   public:
    int eraseOverlapIntervals(std::vector<std::vector<int>>& intervals) {
        int res = 0;
        sort(intervals.begin(), intervals.end(),
             [](const vector<int>& x, const vector<int>& y) { return x[1] < y[1]; });
        int last = intervals[0][1];

        for (int i = 1; i < intervals.size(); i++) {
            if (intervals[i][0] < last) {
                res++;
            } else {
                last = intervals[i][1];
            }
        }

        return res;
    }
};

// Helper test scaffolding
void runTests() {
    Solution sol;
    std::vector<std::vector<std::vector<int>>> inputs = {{{1, 2}, {2, 3}, {3, 4}, {1, 3}},
                                                         {{1, 2}, {1, 2}, {1, 2}},
                                                         {{1, 2}, {2, 3}},
                                                         {{1, 100}, {11, 22}, {1, 11}, {2, 12}},
                                                         {{1, 2}}};
    std::vector<int> expecteds = {1, 2, 0, 2, 0};

    for (size_t i = 0; i < inputs.size(); ++i) {
        int actual = sol.eraseOverlapIntervals(inputs[i]);
        std::cout << actual << " | " << expecteds[i] << std::endl;
    }
}

int main() {
    runTests();
    return 0;
}

File: src/435_non_overlapping_intervals/non_overlapping_intervals.py
from typing import List


class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: x[1])
        res = 0
        last = intervals[0][1]
        for i in range(1, len(intervals)):
            if intervals[i][0] < last:
                res += 1
            else:
                last = intervals[i][1]

        return res


# Helper test scaffolding
def run_tests() -> None:
    cases = [
        ([[1, 2], [2, 3], [3, 4], [1, 3]], 1),
        ([[1, 2], [1, 2], [1, 2]], 2),
        ([[1, 2], [2, 3]], 0),
        ([[1, 100], [11, 22], [1, 11], [2, 12]], 2),
        ([[1, 2]], 0),
    ]

    sol = Solution()
    for intervals, expected in cases:
        actual = sol.eraseOverlapIntervals(intervals)
        print(f"{actual} | {expected}")


if __name__ == "__main__":
    run_tests()
